src/dbnode/storage/mediator.go

```diff
@@ -468,17 +488,29 @@ func (b *mediatorTimeBarrier) maybeRelease() (time.Time, error) {
 	}
 
 	b.mediatorTime = newMediatorTime
+
+	// If all waiters are waiting, we can safely call mutually exclusive external functions.
+	if numWaiters == b.numMaxWaiters {
+		// Drain the channel.
+		for len(b.externalFnCh) > 0 {
+			externalFn := <-b.externalFnCh
+			externalFn()
+		}
```

Comment:
This is racey if there were any other readers on the channel. It is better (and more idiomatic/common) to simply for loop and break when can't read anymore:
```go
for {
  select {
  case fn := <-b.externalFnCh:
    fn()
  default:
    break // Break from loop, no more to read
  }
}
```